src/process_utility.c

```diff
@@ -1209,12 +1238,127 @@ process_index_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
+	Relation chunk_rel;
+
+	Assert(info->extended_options.multitransaction);
+
+	/*
+	 * commit the previous transaction (foreach_chunk requires a transcation
+	 * running, so we cannot commit before this point)
+	 */
+	if (IsTransactionOrTransactionBlock())
+		CommitTransactionCommand();
+
+	/* Start a new transaction for each relation. */
+	StartTransactionCommand();
+	MemoryContextSwitchTo(info->mctx);
+
+#ifdef DEBUG
+	if (info->extended_options.max_chunks == 0)
+		return;
+
+	if (info->extended_options.max_chunks > 0)
+		info->extended_options.max_chunks -= 1;
+
+	if (OidIsValid(info->extended_options.barrier_table))
+	{
+		/*
+		 * For insolation tests, and debugging, it's useful to be able to
+		 * pause CREATE INDEX immediately before it starts working on chunks.
+		 * We acquire and immediately release a lock on a barrier table to do
+		 * this.
+		 */
+		Relation barrier = relation_open(info->extended_options.barrier_table, AccessExclusiveLock);
+
+		relation_close(barrier, AccessExclusiveLock);
+	}
+#endif
+
+	/*
+	 * Change user since chunks are typically located in an internal schema
+	 * and chunk indexes require metadata changes. In the single-tranaction
+	 * case, we do this once for the entire table.
+	 */
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+
+	/*
+	 * Hold a lock on the hypertable, it's index, and the chunk to prevent
+	 * from being altered. Since we use the same relids across transactions,
+	 * there is a potential issue if the id gets reassigned between one
+	 * sub-transaction and the next. CLUSTER has a similar issue.
+	 *
+	 * We grab a ShareLock on the chunk, because that's what CREATE INDEX
+	 * does. For the hypertable, and the hypertable's index, we are ok using
+	 * the weaker AccessShareLock, since we only need to prevent the index
+	 * iteself from being ALTERed or DROPed during this part of index
+	 * creation. TODO: is the above correct?
+	 */
+	hypertable_rel = relation_open(ht->main_table_relid, AccessShareLock);
+	chunk_rel = relation_open(chunk_relid, ShareLock);
+	hypertable_index_rel = relation_open(info->obj.objectId, AccessShareLock);
+
+	chunk = ts_chunk_get_by_relid(chunk_relid, ht->space->num_dimensions, true);
+
+	/*
+	 * use ts_chunk_index_create instead of ts_chunk_index_create_from_stmt to
+	 * handle cases where the index is altered. Validation happens when
+	 * creating the hypertable's index, which goes through the ususal
+	 * DefineIndex mechanism.
+	 */
+	ts_chunk_index_create(hypertable_rel, ht->fd.id, hypertable_index_rel, chunk->fd.id, chunk_rel);
+
+	relation_close(hypertable_index_rel, NoLock);
+	relation_close(chunk_rel, NoLock);
+	relation_close(hypertable_rel, NoLock);
+
+	ts_catalog_restore_user(&sec_ctx);
+}
+
+/*
+ * passing this via static var until we figure out a better way to do it,
+ * maybe reallocate parsetree?
+ */
+
+typedef enum HypertableIndexFlags
+{
+	HypertableIndexFlagMultiTransaction = 0,
+	HypertableIndexFlagBarrierTable,
```

Comment:
definitely should only be used, for some reason I cannot explain I'm a bit concerned with putting `#ifdef TS_DEBUG` right in middle of the definition array (would break our Release mode tests now that I think about it...)

Reply:
let's make this the last element and I think everything should be ok